noether_tpp/src/tool_path_modifiers/organization_modifiers.cpp

```diff
@@ -57,4 +59,175 @@ ToolPaths SnakeOrganizationModifier::modify(ToolPaths tool_paths) const
+    }
+  }
+
+  return segment_order;
+}
+
+/**
+ * @brief Helper function for creating a re-ordered copy of a vector
+ */
+template <typename T>
+static std::vector<T> reorder(const std::vector<T>& input, const std::vector<std::size_t>& order)
+{
+  std::vector<T> output;
+  output.reserve(input.size());
+
+  for (std::size_t idx : order)
+  {
+    output.push_back(input.at(idx));
+  }
+
+  return output;
+}
+
+/**
+ * @brief Organizes the container of tool paths according to the length of all the tool path segments in the tool path
+ * @details The tool paths are sorted from longest to shortest. The organization of the tool path segments within each
+ * tool path and the waypoints within each tool path segment are not changed. The primary intent of this modifier is for
+ * sorting edge tool paths by length.
+ */
+static ToolPaths sortToolPathsByLength(const ToolPaths& tool_paths)
+{
+  std::vector<double> path_lengths;
+  path_lengths.reserve(tool_paths.size());
+
+  for (const ToolPath& path : tool_paths)
+  {
+    double path_length = 0.0;
+
+    for (const ToolPathSegment& segment : path)
+    {
+      for (std::size_t i = 0; i < segment.size() - 1; ++i)
+      {
+        const ToolPathWaypoint& first = segment.at(i);
+        const ToolPathWaypoint& second = segment.at(i + 1);
+        path_length += (first.inverse() * second).translation().norm();
+      }
+    }
+
+    path_lengths.push_back(path_length);
+  }
+
+  // Argsort the path lengths from largest to shortest
+  std::vector<std::size_t> order(path_lengths.size());
+  std::iota(order.begin(), order.end(), 0);
+  std::sort(order.begin(), order.end(), [&path_lengths](const std::size_t& a, const std::size_t& b) {
+    return path_lengths.at(a) > path_lengths.at(b);
+  });
+
+  // Copy the tool paths in the new order
+  return reorder(tool_paths, order);
+}
+
+static Eigen::Vector3d getAverageXAxis(const ToolPathSegment& segment)
+{
+  Eigen::Vector3d mean = Eigen::Vector3d::Zero();
+  for (const ToolPathWaypoint& w : segment)
+  {
+    mean += w.matrix().col(0).head<3>();
+  }
+  return mean / static_cast<double>(segment.size());
+}
+
+StandardEdgePathsOrganizationModifier::StandardEdgePathsOrganizationModifier(const Eigen::Vector3d& start_reference)
+  : start_reference_(start_reference)
+{
+}
+
+ToolPaths StandardEdgePathsOrganizationModifier::modify(ToolPaths tool_paths) const
+{
+  // Sort the waypoints in each tool path segment
+  for (ToolPath& path : tool_paths)
+  {
+    for (ToolPathSegment& segment : path)
+    {
+      // Get the nominal direction of this segment by averaging all the waypoints' x-axes
+      const Eigen::Vector3d ref_dir = getAverageXAxis(segment);
+
+      // Project the position of the waypoints onto the reference direction vector to order them from start to end
+      std::sort(segment.begin(), segment.end(), [&ref_dir](const ToolPathWaypoint& a, const ToolPathWaypoint& b) {
+        return a.translation().dot(ref_dir) < b.translation().dot(ref_dir);
```

Comment:
I can think of a possible failure situation for this modifier as written.  Using the current method of splitting edge paths, the below shape would be broken into four segments as if it were a square - top, right, bottom, left.  However, on the right, merely ordering the waypoints by their position when projected onto the x-axis will cause a number of waypoints to become jumbled.

I do think this function as written, will work perfectly well on raster patterns and their segments.  I just see problems on non-trivial surface shapes, which we do work with on some projects.

My suggestion would be to assume that the edge planner is capable of at least putting out the correct order inside a segment, and focus on re-ordering the `ToolPath` and `ToolPathSegment` objects at the higher levels. If it can't put out correctly-ordered segments, an 'edge planner' is really just an 'edge identifier', returning an unordered set (possibly subset) of edge points.  I feel comfortable setting that bar slightly higher.
```
_______________________
|                      |   _
|                      |_/  |
|                      _____|
|                      |
|                      |
|                      |
|                       \
|_______________________|
```

Reply:
Per our discussion earlier, we expect the planner itself to order the waypoints correctly so this part of the function will be deleted to avoid scenarios like the one you pointed out